ambry-clustermap/src/main/java/com.github.ambry.clustermap/CloudServiceClusterChangeHandler.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2020 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.clustermap;
+
+import com.github.ambry.clustermap.HelixClusterManager.ClusterChangeHandlerCallback;
+import com.github.ambry.config.ClusterMapConfig;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Currently, this just adds a single {@link ReplicaType#CLOUD_BACKED} replica every time a partition is added to allow
+ * for requests to be routed to cloud services. Currently it is only registered as a {@link ClusterMapChangeListener},
+ * but once the clustermap natively supports VCR clusters, this class can implement {@link HelixAwareClusterChangeHandler} and
+ * listen for changes in the VCR cluster.
+ */
+class CloudServiceClusterChangeHandler implements ClusterMapChangeListener, ClusterChangeHandler {
+  private static final Logger logger = LoggerFactory.getLogger(CloudServiceClusterChangeHandler.class);
+  private final Set<String> knownPartitions = ConcurrentHashMap.newKeySet();
+  private final ConcurrentHashMap<String, CloudServiceReplica> partitionToReplica = new ConcurrentHashMap<>();
+  private final List<ClusterMapChangeListener> listeners = new CopyOnWriteArrayList<>();
+  private final AtomicLong errorCount = new AtomicLong(0);
+  private final String dcName;
+  private final ClusterMapConfig clusterMapConfig;
+  private final ClusterChangeHandlerCallback changeHandlerCallback;
+  private final CloudServiceDataNode cloudServiceDataNode;
+  private final String cloudServiceInstanceName;
+
+  public CloudServiceClusterChangeHandler(String dcName, ClusterMapConfig clusterMapConfig,
+      ClusterChangeHandlerCallback changeHandlerCallback) throws Exception {
+    this.dcName = dcName;
+    this.clusterMapConfig = clusterMapConfig;
+    this.changeHandlerCallback = changeHandlerCallback;
+    cloudServiceDataNode = new CloudServiceDataNode(dcName, clusterMapConfig);
+    cloudServiceInstanceName =
+        ClusterMapUtils.getInstanceName(cloudServiceDataNode.getDatacenterName(), cloudServiceDataNode.getPort());
```

Comment:
I tried this. The problem I ran into is that the method `ClusterMap.getDataNodeId(...)` requires an int port number to be passed in and generates the instance name before calling `ClusterChangeHandler.getDataNode`.

The options would be:
1. Keep it as it is with a -1 port for CloudServiceDataNodes
2. Don't support `getDataNode` in `CloudServiceClusterChangeHandler`. This could be a possibility since it probably has very limited value in this case.
3. Change `ClusterMap.getDataNodeId` method to take a nullable `Integer` instead of `int`

Which do you prefer, @jsjtzyy?

Reply:
If `getDataNode` is unlikely to be used in `CloudServiceClusterChangeHandler`, yeah, we don't have to support it. (Let's go with option 2, what do you think?)

Reply:
implemented 2